Generate MVP wasm for fuzzing

jfoote opened the pull request:
Hello,

I've developed a [cargo-fuzz WebAssembly fuzz target for cretonne](https://github.com/cretonne/cretonne/pull/306) that indirectly calls the binaryen-rs `translateToFuzz` function. Because cretonne does not support post-MVP WebAssembly functionality like atomics, I implemented a simple hack in binaryen-rs (in this PR) to restrict generation to only MVP WebAssembly. I don't think what I've done here is the correct approach, I just wanted to supply the code for reference.

Do you have any advice or guidance on how you'd like to see something like this implemented?

Thanks,
Jon

pepyakin commented:
Ideally, the best solution is to have [means of configuring](https://github.com/WebAssembly/binaryen/blob/fd3b3e54bd97abbf8269b33d937ad2f44ba4bb60/src/tools/fuzzing.h#L178) `TranslateToFuzzReader`, but this will definitely will require to change binaryen. Workaround to validate will also work. 

If you would like to go with validate workaround, the thing I would like to be changed is to move validation outside the `translateToFuzz` shim function. Instead, we could add overload of [`is_valid`](https://github.com/pepyakin/binaryen-rs/blob/6ab57d19a82cda0cd9174b42655a22120f333e0d/src/lib.rs#L74) to actually take param that will specify the feature set to validate against. Since binaryen-c API also [doesn't support](https://github.com/WebAssembly/binaryen/blob/fd3b3e54bd97abbf8269b33d937ad2f44ba4bb60/src/binaryen-c.cpp#L2004) setting we will need to update the shim or, again, make PR to binaryen.

jfoote commented:
Hello,

binaryen accepted a [patch](https://github.com/WebAssembly/binaryen/pull/1513) that makes emitting atomics optional for the translate-to-fuzz feature. I am not sure if this is what you had in mind exactly, but I updated the underlying branch here to implement a corresponding `translate_to_fuzz_mvp` function in binaryen-rs. 

When you have a moment could you let me know what you think?

Note that I implemented this against upstream binaryen rather than your [fork](https://github.com/pepyakin/binaryen/tree/d12b351bea74b93501328a52c07e780b0ed78086) (to include the ttf MVP-only patch) -- if your fork is updated I think it should work as expected.

Thanks,
Jon

jfoote commented:
Hello @pepyakin,

I updated the PR branch to use the most recent version of binaryen/master to verify that binaryen-rs will work as expected. It works for me locally and passes all integration tests locally -- I am not sure why travis is failing (looking at the log it may be a travis container resource issue, but I am unable to restart the build / debug). Feel free to switch the submodule back at your convienence ofc (or direct me to when you are ready).

Regardless, could you please review this when you get a chance?

Thanks,
Jon

pepyakin commented:
Hey, yeah sure. Ill take a look as soon as I get back from holidays!

> On 4 May 2018, at 16:34, Jonathan Foote <notifications@github.com> wrote:
> 
> Hello @pepyakin,
> 
> I updated the PR branch to use the most recent version of binaryen/master to verify that binaryen-rs will work as expected. It works for me locally and passes all integration tests locally -- I am not sure why travis is failing (looking at the log it may be a travis container resource issue, but I am unable to restart the build / debug). Feel free to switch the submodule back at your convienence ofc (or direct me to when you are ready).
> 
> Regardless, could you please review this when you get a chance?
> 
> Thanks,
> Jon
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jfoote commented:
Thanks, enjoy!

pepyakin commented on binaryen-sys/Shim.cpp:
```diff
@@ -8,17 +8,20 @@
 using namespace wasm;
 
-extern "C" BinaryenModuleRef translateToFuzz(const char *data, size_t len) {
+extern "C" BinaryenModuleRef translateToFuzz(const char *data, size_t len, bool emitAtomics = true) {
```
I feel a little wary about making this param with a default value simply because I don't understand how `extern "C"` works with arguments with default values. At the same time default value isn't needed (right?) here since this function is only supposed to be called via FFI. 
With all that, I think that it would be better to just remove the default value from the argument.

jfoote replied:
Good point -- removed

pepyakin commented on src/tools.rs:
```diff
@@ -29,8 +46,21 @@ mod tests {
+        for _ in 0..1000 {
+            let mut rng = rand::thread_rng();
+            rng.fill_bytes(&mut seed);
+            let module = translate_to_fuzz(&seed);
```
I guess you meant to use `translate_to_fuzz_mvp`, right?

jfoote replied:
Oops, yes -- fixed

jfoote commented:
@pepyakin thank you for the feedback! I fixed the bugs you caught and merged in #5, I think this should be good to go.

pepyakin commented:
Great job, thank you!
